### PR TITLE
DM-42945: Implement RemoteButler.exists

### PR DESCRIPTION
**Pull request by dhirving**
Implemented `RemoteButler.exists()` and `RemoteButler._exists_many()`.  To get test coverage for these methods, expanded the set of tests run against Butler server to include those from `FileDatastoreButlerTests`, and moved `testPruneDatasets` from `PosixDatastoreButlerTestCase` to `FileDatastoreButlerTests`.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/daf/butler/remote_butler/server_models.py [hunk 1]
```diff
@@ -64,4 +64,17 @@ class GetFileByDataIdRequestModel(pydantic.BaseModel):
     collections: CollectionList
 
 
-GetFileResponseModel = FileDatastoreGetPayload
+class GetFileResponseModel(pydantic.BaseModel):
+    """Response model for get_file and get_file_by_data_id."""
+
+    dataset_ref: SerializedDatasetRef
+    artifact: FileDatastoreGetPayload | None
```
Comment (dhirving) [inline python/lsst/daf/butler/remote_butler/server_models.py:71]: This change breaks compatibility with previous versions of client code. Discussed with Russ and we are just going to bump the Butler library versions for datalinker/vo-cutouts and update Phalanx simultaneously with the new version of Butler server.

File: python/lsst/daf/butler/remote_butler/_remote_butler.py [hunk 2]
```diff
@@ -439,8 +461,7 @@ def _exists_many(
         *,
         full_check: bool = True,
     ) -> dict[DatasetRef, DatasetExistence]:
-        # Docstring inherited.
-        raise NotImplementedError()
+        return {ref: self.exists(ref, full_check=full_check) for ref in refs}
```
Comment (dhirving) [inline python/lsst/daf/butler/remote_butler/_remote_butler.py:464]: This is obviously pretty inefficient. (Not radically worse than the DirectButler version, though -- it also does a getDataset query per-ref.)  I anticipate modifying the file info endpoint to support batching to implement `get_many_uris`, which could also be used to get a speedup here.

File: tests/test_butler.py [hunk 3]
```diff
@@ -1510,68 +1514,21 @@ def testRemoveRuns(self) -> None:
             butler.registry.removeDatasetType(("test*", "test*"))
         self.assertIn("not defined", "\n".join(cm.output))
 
+    def remove_dataset_out_of_band(self, butler: Butler, ref: DatasetRef) -> None:
+        """Simulate an external actor removing a file outside of Butler's
+        knowledge.
 
-class PosixDatastoreButlerTestCase(FileDatastoreButlerTests, unittest.TestCase):
```
Comment (dhirving) [inline tests/test_butler.py:1514]: This diff reads a little weird.  What actually happened is that `testPruneDatasets` moved up from `PosixDatastoreButlerTestCase` to `FileDatastoreButlerTests`, but apparently git feels the diff is smaller to move this chunk down below `testPruneDatasets`.

File: python/lsst/daf/butler/remote_butler/_remote_butler.py [hunk 4]
```diff
@@ -716,6 +737,14 @@ def _extract_dataset_type(datasetRefOrType: DatasetRef | DatasetType | str) -> D
         return None
 
 
+def _to_file_payload(get_file_response: GetFileResponseModel) -> FileDatastoreGetPayload:
+    if get_file_response.artifact is None:
+        ref = get_file_response.dataset_ref
+        raise FileNotFoundError(f"Dataset is known, but artifact is not available. (datasetId='{ref.id}')")
```
Comment (dhirving) [inline python/lsst/daf/butler/remote_butler/_remote_butler.py:743]: These lines will gain coverage in the next ticket; some of the unit tests for getDeferred do the necessary twiddling of the datastore records.

File: tests/test_butler.py (hunk 3 above)
Comment (timj) [reply tests/test_butler.py:1514]: difftastic did a pretty good job of spotting the move.

**Review by timj — APPROVED** [review]
Looks good.

**Pull request merged**